# Week 10 "Pre-lecture" HW07

### 1. Explain succinctly in your own words (but working with a ChatBot if needed)...<br>

1. the difference between **Simple Linear Regression** and **Multiple Linear Regression**; and the benefit the latter provides over the former
- Simple Linear Regression models the relationship between a single predictor and an outcome, using an equation like:
  $$
  \text{outcome} = \beta_0 + \beta_1 \cdot \text{predictor}
  $$
- Multiple Linear Regression expands this by incorporating multiple predictors, which can capture more complex relationships. Its form is:
  $$
  \text{outcome} = \beta_0 + \beta_1 \cdot \text{predictor}_1 + \beta_2 \cdot \text{predictor}_2 + \dots + \beta_n \cdot \text{predictor}_n
  $$
- Benefit: Multiple Linear Regression allows for greater accuracy and flexibility by accounting for multiple factors that may influence the outcome, rather than just one.

---

2. the difference between using a **continuous variable** and an **indicator variable** in **Simple Linear Regression**; and these two **linear forms**
- A continuous variable takes on any value within a range and is used in regression directly. The equation form remains:
  $$
  \text{outcome} = \beta_0 + \beta_1 \cdot \text{continuous predictor}
  $$
- An indicator variable represents categorical data with binary values, usually 0 or 1, to indicate the presence or absence of a category. It modifies the model form to:
  $$
  \text{outcome} = \beta_0 + \beta_1 \cdot 1(\text{indicator})
  $$
- Linear Forms: Continuous variables model linear effects of gradually increasing or decreasing predictor values, while indicator variables add a fixed effect when a specific category is present.

---

3. the change that happens in the behavior of the model (i.e., the expected nature of the data it models) when a single **indicator variable** is introduced alongside a **continuous variable** to create a **Multiple Linear Regression**; and these two **linear forms** (i.e., the **Simple Linear Regression** versus the **Multiple Linear Regression**)
- When adding an indicator variable to a Simple Linear Regression with a continuous variable, we transition to a Multiple Linear Regression model, as both variables now jointly predict the outcome:
  $$
  \text{outcome} = \beta_0 + \beta_1 \cdot \text{continuous predictor} + \beta_2 \cdot 1(\text{indicator})
  $$
- Model Behavior: The indicator variable shifts the outcome based on category (0 or 1), while the continuous variable maintains its linear influence across all observations. This setup creates two parallel lines, one for each category of the indicator, but with the same slope for the continuous variable.

---

4. the effect of adding an **interaction** between a **continuous** and an **indicator variable** in **Multiple Linear Regression** models; and this **linear form**
- Adding an interaction term between a continuous and an indicator variable allows for different slopes for each category of the indicator:
  $$
  \text{outcome} = \beta_0 + \beta_1 \cdot \text{continuous predictor} + \beta_2 \cdot 1(\text{indicator}) + \beta_3 \cdot \text{continuous predictor} \cdot 1(\text{indicator})
  $$
- Effect: This changes the behavior so each category (0 or 1 in the indicator variable) has its own slope for the continuous predictor, allowing the relationship between the predictor and the outcome to vary across categories. This captures more nuanced relationships between variables.

---

5. the behavior of a **Multiple Linear Regression** model (i.e., the expected nature of the data it models) based only on **indicator variables** derived from a **non-binary categorical variable**; this **linear form**; and the necessarily resulting **binary variable encodings** it utilizes

- When encoding a non-binary categorical variable (e.g., “color” with categories red, blue, and green), we need indicator variables for each category except one baseline (often the most common category). For instance, if we choose "green" as the baseline, we would set up the model as follows:
  $$
  \text{outcome} = \beta_0 + \beta_1 \cdot 1(\text{red}) + \beta_2 \cdot 1(\text{blue})
  $$
  where:
  - $\beta_0$ represents the expected outcome for the baseline category (green),
  - $\beta_1$ represents the difference in outcome between "red" and "green,"
  - $\beta_2$ represents the difference in outcome between "blue" and "green."

- Model Behavior: This Multiple Linear Regression model assumes that each category within the categorical variable (e.g., color) has a distinct, fixed impact on the outcome, relative to the baseline. The model's outcome varies across categories but remains constant within each category. This creates a set of parallel lines on the outcome axis, one for each category, where only the intercept differs according to the specific category.

- Linear Form: In this linear form, each indicator variable represents a category, and only one indicator variable will be active (equal to 1) for any given observation, while the others are 0. This structure allows the model to differentiate the effect of each category relative to the baseline. 

- Binary Variable Encodings: For a categorical variable with $k$ categories, we use $k-1$ indicator variables to encode all categories except one. This ensures that each category is compared to the baseline (here, "green"). The model's intercept $\beta_0$ represents the baseline category, and each $\beta$ coefficient shows the expected change in the outcome from the baseline when the observation falls within a specific category.

---
### Summaries of my ChatBot session (Q#1) (01)
**Link to the this chat log history for "Pre-lecture" HW:** https://chatgpt.com/share/67316d43-8ecc-8011-a955-39389d1c332b

**1. Simple Linear Regression vs. Multiple Linear Regression**  
   - **Simple Linear Regression** uses one predictor variable to model a dependent variable, resulting in a linear relationship between the two.
   - **Multiple Linear Regression** involves multiple predictors, allowing the model to account for additional complexity and potentially improve prediction accuracy by incorporating more variables.
   
**2. Continuous vs. Indicator Variable in Simple Linear Regression**  
   - **Continuous variable**: Takes on any numeric value within a range, allowing the model to create a linear trend.
   - **Indicator variable**: A binary (0 or 1) variable representing categories, creating distinct intercepts for each group but not a continuous trend.
   - **Linear Forms**: A continuous variable allows for a linear relationship with a slope, while an indicator variable only shifts the intercept, creating distinct levels for each category.

**3. Effect of Adding an Indicator Variable to Create Multiple Linear Regression**  
   - Adding an indicator variable along with a continuous variable allows for separate intercepts for each category while sharing the same slope, modeling different starting points without changing the trend.
   - **Linear Forms**: In Simple Linear Regression, we have a single slope. In Multiple Linear Regression with an indicator, the intercept adjusts by the indicator’s coefficient, creating parallel lines for each category.

**4. Interaction Between Continuous and Indicator Variables in Multiple Linear Regression**  
   - Introducing an **interaction term** between a continuous and an indicator variable allows the slope for each category to differ, making the model more flexible and able to capture unique trends within each group.
   - **Linear Form**: Each category has a distinct slope and intercept when the interaction term is included, allowing for varied relationships between the continuous variable and the dependent variable.

**5. Multiple Linear Regression with Only Indicator Variables for Non-Binary Categorical Variables**  
   - When only indicator variables are used for a categorical variable with multiple levels, each category has its own intercept, but there is no continuous trend.
   - **Linear Form**: Using \( k-1 \) indicators for \( k \) categories creates a model with binary encodings, shifting the intercept based on category membership.

---
### 2. Explain in your own words (but working with a ChatBot if needed) what the specific (outcome and predictor) variables are for the scenario below; whether or not any meaningful interactions might need to be taken into account when predicting the outcome; and provide the linear forms with and without the potential interactions that might need to be considered<br>

> Imagine a company that sells sports equipment. The company runs advertising campaigns on TV and online platforms. The effectiveness of the TV ad might depend on the amount spent on online advertising and vice versa, leading to an interaction effect between the two advertising mediums.    

#### 1. Explain how to use these two formulas to make **predictions** of the **outcome**, and give a high level explaination in general terms of the difference between **predictions** from the models with and without the **interaction** 
To make predictions, plug the values for $TV$ budget and $Online$ budget into the linear equations. There are two versions of the model:

- Without Interaction:
  $$
  \text{effectiveness} = \beta_0 + \beta_{\text{TV}} \cdot \text{TV budget} + \beta_{\text{Online}} \cdot \text{Online budget}
  $$
  In this model, each budget independently contributes to the predicted outcome. Increasing the $TV$ budget by a certain amount results in a fixed increase in effectiveness, regardless of the $Online$ budget. Similarly, increasing the $Online$ budget results in a fixed increase in effectiveness, regardless of the $TV$ budget.

- With Interaction:
  $$
  \text{effectiveness} = \beta_0 + \beta_{\text{TV}} \cdot \text{TV budget} + \beta_{\text{Online}} \cdot \text{Online budget} + \beta_{\text{TV-Online}} \cdot (\text{TV budget} \times \text{Online budget})
  $$
  This model includes an interaction term, $\beta_{\text{TV-Online}}$, which captures the combined effect of both budgets. When both the $TV$ and $Online$ budgets are high, this term amplifies the predicted outcome, indicating a synergistic relationship where the combined effect of both budgets is greater than the sum of their separate contributions.

- General Difference: In the additive model (without interaction), the impact of each advertising type is independent, so the effect of increasing one budget doesn’t depend on the level of the other. In the synergistic model (with interaction), the outcome prediction includes a boost when both budgets are high, allowing for a potentially amplified effect that reflects synergy between the two advertising mediums.


#### 2. Explain how to update and use the implied two formulas to make predictions of the outcome if, rather than considering two continuous predictor variables, we instead suppose the advertisement budgets are simply categorized as either "high" or "low" (binary variables)
If we categorize each advertising budget as either “high” or “low” (binary indicators), we update the model as follows:

- Without Interaction (Binary Indicators):
  $$
  \text{effectiveness} = \beta_0 + \beta_{\text{TV}} \cdot 1(\text{TV = high}) + \beta_{\text{Online}} \cdot 1(\text{Online = high})
  $$
  In this form, $1(\text{TV = high})$ and $1(\text{Online = high})$ are binary indicators that take a value of 1 if the budget is “high” and 0 if it is “low.” Each budget’s effect on effectiveness is independent, so the effect of high $TV$ spending remains the same whether the online budget is high or low, and vice versa.

- With Interaction (Binary Indicators):
  $$
  \text{effectiveness} = \beta_0 + \beta_{\text{TV}} \cdot 1(\text{TV = high}) + \beta_{\text{Online}} \cdot 1(\text{Online = high}) + \beta_{\text{TV-Online}} \cdot (1(\text{TV = high}) \times 1(\text{Online = high}))
  $$
  Here, the interaction term $\beta_{\text{TV-Online}}$ accounts for a synergistic effect when both advertising budgets are high. If both budgets are high, the predicted effectiveness includes an additional boost, indicating that the impact is “more than just the sum of its parts.”

- Using the Updated Models for Predictions: 
  - Additive Model (Without Interaction): Simply add the effects of high $TV$ and online budgets independently. The outcome reflects a straightforward combination of the two budgets, without any added synergy.
  - Synergistic Model (With Interaction): When both budgets are high, the prediction includes an additional interaction effect, indicating an enhanced effectiveness. This model is suitable if the combined high spending on both platforms creates a stronger impact on the outcome than each individually.


---
### Summaries of my ChatBot session (Q#2) (02)
**Link to the this chat log history for "Pre-lecture" HW:** https://chatgpt.com/share/673236c6-ebf4-8011-8d9e-e8ec64720ca2

**1. Outcome and Predictor Variables**

- **Outcome Variable**: Sales or conversion rate resulting from the advertising campaigns (dependent variable).
- **Predictor Variables**:
  - **TV Advertising Spend**: Amount spent on TV advertising.
  - **Online Advertising Spend**: Amount spent on online advertising.

The goal is to determine if these predictor variables have an additive or synergistic effect on the outcome.

**2. Interaction Consideration**

There may be an **interaction effect** between TV and online advertising. If this interaction exists, the effect of TV spending on sales could depend on the amount spent online, and vice versa. In such a case, the combined effect of TV and online ads would not simply be additive but synergistic, meaning the impact is more than just the sum of individual contributions.

**3. Linear Model Forms**

- **Without Interaction**:
  \[
  \text{Sales} = \beta_0 + \beta_1 \times \text{TV} + \beta_2 \times \text{Online}
  \]
  - **Interpretation**: Assumes an additive effect. The influence of TV and online spending on sales is independent of each other.

- **With Interaction**:
  \[
  \text{Sales} = \beta_0 + \beta_1 \times \text{TV} + \beta_2 \times \text{Online} + \beta_3 \times (\text{TV} \times \text{Online})
  \]
  - **Interpretation**: Assumes a synergistic effect. The impact of one type of ad spending on sales depends on the level of the other, leading to a “more than the sum of its parts” outcome when both are high.

**4. Using Models for Prediction and Key Differences**

- **Additive Model (No Interaction)**: Predicts sales by substituting the values of TV and online spending, assuming each variable independently affects sales.
- **Synergistic Model (With Interaction)**: Predicts sales by substituting values of TV and online spending, factoring in the interaction term. The effect of one advertising type changes based on the level of the other.

**5. Binary Predictor Variables (High vs. Low Spending)**

If we categorize spending as "high" or "low" instead of using continuous values, the models become:

- **Without Interaction**:
  \[
  \text{Sales} = \beta_0 + \beta_1 \times \text{TV\_High} + \beta_2 \times \text{Online\_High}
  \]
  - **Interpretation**: Predicts that high TV or high online spending each independently influences sales in an additive way.

- **With Interaction**:
  \[
  \text{Sales} = \beta_0 + \beta_1 \times \text{TV\_High} + \beta_2 \times \text{Online\_High} + \beta_3 \times (\text{TV\_High} \times \text{Online\_High})
  \]
  - **Interpretation**: Allows for a synergistic effect when both TV and online spending are high, meaning their combined effect is greater than their individual effects alone.

---

**Overall Summary**: The additive models assume that each form of advertising independently contributes to sales, while the synergistic models incorporate an interaction effect, where the impact of one form of advertising depends on the level of the other. This interaction can be modeled both with continuous and binary variables, offering insights into how the combination of TV and online ad spend influences sales outcomes. 


### 4. Explain the apparent contradiction between the factual statements regarding the fit below that "the model only explains 17.6% of the variability in the data" while at the same time "many of the *coefficients* are larger than 10 while having *strong* or *very strong evidence against* the *null hypothesis* of 'no effect'"


In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [2]:
import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        21:30:05   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In this situation, two different ways of looking at a model might seem to contradict each other at first glance. On one hand, we’re told that the model only explains 17.6% of the variability in the data. This means that the $ R^2 $ value, which shows how well the model accounts for changes in the outcome variable $ y $, is relatively low. Typically, a low $ R^2 $ suggests that the model doesn’t do a great job of explaining why $ y $ varies as it does. In other words, most of the factors that influence $ y $ are still unknown or unexplained by the variables included in this model. 

However, at the same time, we see that many of the model’s coefficients are large and have very small p-values, meaning there’s strong or very strong evidence against the idea that these predictors have “no effect” on $ y $. Statistically, a small p-value (below 0.05, for instance) tells us that it’s very unlikely for the observed relationship between a predictor and $ y $ to be due to random chance alone. So, even though the model doesn’t explain much of the variation in $ y $, these individual predictors do appear to have real, non-random relationships with $ y $. This can feel like a contradiction: how can predictors be significant if the model as a whole doesn’t explain much?

To understand this, it’s important to realize that $ R^2 $ and p-values are telling us different things. $ R^2 $ is about how much of the overall “story” behind $ y $’s variability our model captures. A low $ R^2 $ means we’re missing large parts of this story—there are other, potentially unknown factors influencing $ y $ that we haven’t included in the model. In contrast, p-values focus on whether each predictor, by itself, has a meaningful and consistent effect on $ y $. A small p-value for a predictor tells us that, despite other unexplained factors, this predictor still shows a clear, non-random relationship with $ y $ when we control for the other predictors. In other words, each predictor’s effect on $ y $ is real and statistically significant, but the total contribution of these predictors doesn’t explain most of $ y $’s variability.

So, in short, the model’s low $ R^2 $ tells us that there’s still a lot about $ y $ that our predictors don’t account for, while the small p-values tell us that the relationships we do see are statistically significant. The predictors we’ve included might influence $ y $ in a meaningful way, but they don’t capture most of the factors driving $ y $’s changes. This isn’t a contradiction; it just highlights the difference between measuring a model’s overall fit and identifying specific, real relationships between predictors and the outcome.


# Week 10 "Potr-lecture" HW07

### 5. Discuss the following (five cells of) code and results with a ChatBot and based on the understanding you arrive at in this conversation explain what the following (five cells of) are illustrating

In [3]:
import numpy as np
from sklearn.model_selection import train_test_split

fifty_fifty_split_size = int(pokeaman.shape[0]*0.5)

# Replace "NaN" (in the "Type 2" column with "None")
pokeaman.fillna('None', inplace=True)

np.random.seed(130)
pokeaman_train,pokeaman_test = \
  train_test_split(pokeaman, train_size=fifty_fifty_split_size)
pokeaman_train

#                     Name   Type 1   Type 2  HP  Attack  Defense  \
370  338                  Solrock     Rock  Psychic  70      95       85   
6      6                Charizard     Fire   Flying  78      84       78   
242  224                Octillery    Water     None  75     105       75   
661  600                    Klang    Steel     None  60      80       95   
288  265                  Wurmple      Bug     None  45      45       35   
..   ...                      ...      ...      ...  ..     ...      ...   
522  471                  Glaceon      Ice     None  65      60      110   
243  225                 Delibird      Ice   Flying  45      55       45   
797  720      HoopaHoopa Confined  Psychic    Ghost  80     110       60   
117  109                  Koffing   Poison     None  40      65       95   
409  373  SalamenceMega Salamence   Dragon   Flying  95     145      130   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
370       55       65     70           3      False  
6        109       85    100           1      False  
242      105       75     45           2      False  
661       70       85     50           5      False  
288       20       30     20           3      False  
..       ...      ...    ...         ...        ...  
522      130       95     65           4      False  
243       65       45     75           2      False  
797      150      130     70           6       True  
117       60       45     35           1      False  
409      120       90    120           3      False  

[400 rows x 12 columns]

This cell is setting up the data to help us test how well a model we train can work on new, unseen data. By splitting the dataset in half, we’ll train the model on one half (the "training" data) and then check its performance on the other half (the "testing" data). This split is key because it lets us see if the model is just memorizing the training data or if it’s able to make good predictions on data it hasn't seen before.

The reason we’re doing this is to measure the "in-sample" versus "out-of-sample" performance of the model. If the model performs well on both the training and testing data, it shows that it's generalizing well. But if it only does well on the training data and poorly on the testing data, it might be overfitting—meaning it learned patterns specific to the training set that don’t hold up elsewhere. So, this cell is really about setting up a way to test whether our model’s predictions are actually useful beyond the data it was trained on.

---

In [4]:
model_spec3 = smf.ols(formula='HP ~ Attack + Defense', 
                      data=pokeaman_train)
model3_fit = model_spec3.fit()
model3_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        21:30:05   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

This cell is about building and analyzing a linear regression model to see if we can predict the "HP" of Pokémon based on their "Attack" and "Defense" stats. Here, we’re only using the training data (the first half of the dataset) to create this model. The ols function sets up the model with "HP" as the outcome variable we want to predict, and "Attack" and "Defense" as the predictor variables that might help explain variations in "HP."

When we run model3_fit.summary(), it gives us detailed information about how well this model fits the training data, including key metrics like R-squared, coefficients, and p-values. The R-squared here tells us how much of the variation in HP can be explained by Attack and Defense stats in this training dataset, which is crucial to understanding in-sample performance. This output helps us see if the relationship we've modeled seems strong and if the predictor variables are actually meaningful for predicting HP.

This cell is all about creating a model on our training data and then evaluating its initial fit before we even look at how well it might perform on new, unseen data.

---

In [5]:
yhat_model3 = model3_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model3_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model3)[0,1]**2)

'In sample' R-squared:     0.14771558304519894
'Out of sample' R-squared: 0.21208501873920738


This cell is checking how well the model we trained can predict HP for new data it hasn’t seen before. First, it uses the model to predict HP values for the testing data—the half of the dataset that wasn’t used to train the model. These predictions (yhat_model3) are the model’s best guesses for HP based on Attack and Defense stats in the testing set. At the same time, y holds the actual HP values from this testing set so we can compare the predictions to the real values.

To see how well the model does, we look at two R-squared values: the "in-sample" R-squared and the "out-of-sample" R-squared. The in-sample R-squared tells us how well the model fits the training data, which it was built to match. The out-of-sample R-squared, on the other hand, shows how well the model’s predictions match actual HP values in the testing set, where it’s predicting HP for data it hasn’t seen.

By comparing these two values, we can check if the model is overfitting. If the out-of-sample R-squared is much lower than the in-sample R-squared, it suggests that the model might be memorizing patterns in the training data rather than learning relationships that work on new data. This comparison shows us whether our model is generalizing well to new data or just fitting the training data closely without broader accuracy.

---

In [6]:
model4_linear_form = 'HP ~ Attack * Defense * Speed * Legendary'
model4_linear_form += ' * Q("Sp. Def") * Q("Sp. Atk")'
# DO NOT try adding '* C(Generation) * C(Q("Type 1")) * C(Q("Type 2"))'
# That's 6*18*19 = 6*18*19 possible interaction combinations...
# ...a huge number that will blow up your computer

model4_spec = smf.ols(formula=model4_linear_form, data=pokeaman_train)
model4_fit = model4_spec.fit()
model4_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.467
Model:                            OLS   Adj. R-squared:                  0.369
Method:                 Least Squares   F-statistic:                     4.764
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           4.23e-21
Time:                        21:30:05   Log-Likelihood:                -1738.6
No. Observations:                 400   AIC:                             3603.
Df Residuals:                     337   BIC:                             3855.
Df Model:                          62                                         
Covariance Type:            nonrobust                                         
====================================================================================================================================
                                                                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------------
Intercept                                                          521.5715    130.273      4.004      0.000     265.322     777.821
Legendary[T.True]                                                   -6.1179      2.846     -2.150      0.032     -11.716      -0.520
Attack                                                              -8.1938      2.329     -3.518      0.000     -12.775      -3.612
Attack:Legendary[T.True]                                         -1224.9610    545.105     -2.247      0.025   -2297.199    -152.723
Defense                                                             -6.1989      2.174     -2.851      0.005     -10.475      -1.923
Defense:Legendary[T.True]                                         -102.4030     96.565     -1.060      0.290    -292.350      87.544
Attack:Defense                                                       0.0985      0.033      2.982      0.003       0.034       0.164
Attack:Defense:Legendary[T.True]                                    14.6361      6.267      2.336      0.020       2.310      26.963
Speed                                                               -7.2261      2.178     -3.318      0.001     -11.511      -2.942
Speed:Legendary[T.True]                                            704.8798    337.855      2.086      0.038      40.309    1369.450
Attack:Speed                                                         0.1264      0.038      3.351      0.001       0.052       0.201
Attack:Speed:Legendary[T.True]                                       5.8648      2.692      2.179      0.030       0.570      11.160
Defense:Speed                                                        0.1026      0.039      2.634      0.009       0.026       0.179
Defense:Speed:Legendary[T.True]                                     -6.9266      3.465     -1.999      0.046     -13.742      -0.111
Attack:Defense:Speed                                                -0.0016      0.001     -2.837      0.005      -0.003      -0.001
Attack:Defense:Speed:Legendary[T.True]                              -0.0743      0.030     -2.477      0.014      -0.133      -0.015
Q("Sp. Def")                                                        -5.3982      1.938     -2.785      0.006      -9.211      -1.586
Legendary[T.True]:Q("Sp. Def")                                    -282.2496    126.835     -2.225      0.027    -531.738     -32.761
Attack:Q("Sp. Def")                                                  0.1094      0.034      3.233      0.001       0.043       0.176
Attack:Legendary[T.True]:Q("Sp. Def")                               12.6503      5.851      2.162      0.031       1.141      24.160
Defense:Q("Sp. Def")    

This cell is creating a more complex model to predict HP by including multiple features and their interactions. Here, the formula used in the model includes Attack, Defense, Speed, Legendary, Sp. Def, and Sp. Atk, along with all possible interaction terms among these variables. By using interactions, the model can capture how combinations of these attributes might jointly affect HP rather than just looking at each variable’s isolated effect.

Once defined, this model is fit to the training data (pokeaman_train), and model4_fit.summary() is used to display a summary of the model’s performance on this data. This output includes R-squared, coefficients, and statistical significance for each term, which helps us understand whether the added complexity improves the model’s fit to the training data.

Overall, this cell is illustrating a move toward a higher-dimensional model, capturing more detailed relationships between the variables in the hopes of improving prediction accuracy.

---

In [7]:
yhat_model4 = model4_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model4_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model4)[0,1]**2)

'In sample' R-squared:     0.46709442115833855
'Out of sample' R-squared: 0.002485342598992873


This cell uses the same general approach as the third cell, where we compare in-sample and out-of-sample R-squared values to assess model performance and generalizability. Both cells aim to determine how well each model (first the simpler model in the third cell, then the more complex model in this cell) fits the training data and, importantly, whether it can generalize to predict HP accurately on the testing data.

---
### Summaries of my ChatBot session (Q#5) (03)
**Link to the this chat log history for "Pre-lecture" HW:** https://chatgpt.com/share/6732ede8-8320-8011-bdc2-df917da57f51

Here's a structured summary of our discussion on evaluating model performance and generalizability through in-sample and out-of-sample R-squared values:

#### Cell 1
**Code Explanation:**
We began by examining the code that splits a dataset (`pokeaman`) into training and testing sets, enabling us to evaluate the model's generalizability. Specifically:
- `train_test_split` was used to divide the data, with a 50/50 split.
- Missing values in the "Type 2" column were replaced with `'None'`, ensuring data consistency.
- Setting a random seed guaranteed reproducibility of the split.

**Purpose:**
This split sets up a foundation for comparing the "in-sample" model performance (training set) against the "out-of-sample" performance (test set), crucial for assessing overfitting and generalizability.

#### Cell 2
**Code Explanation:**
Next, we explored the fitting of a **simple linear regression model** using Ordinary Least Squares (OLS) with `HP` as the dependent variable and `Attack` and `Defense` as predictors.
- We noted that this initial model provides an **in-sample R-squared** value, showing how much of the variance in `HP` is explained within the training set.

**Purpose:**
This model gives a baseline in-sample performance metric. We discussed how the R-squared value here only reflects the fit on the training data, which might not hold for new data.

#### Cell 3
**Code Explanation:**
We calculated both **in-sample** and **out-of-sample** R-squared values for the simple model by:
- Comparing the actual `HP` values in the test set with predicted values from the model.
- Printing both in-sample R-squared (from the training set) and out-of-sample R-squared (based on test data predictions).

**Purpose:**
This comparison lets us evaluate generalizability: if out-of-sample R-squared is much lower than in-sample, the model may be overfitting to the training data. A close match between these values suggests better generalizability.

#### Cell 4
**Code Explanation:**
We expanded the model with a more complex formula, adding predictors (`Speed`, `Legendary`, `Sp. Def`, `Sp. Atk`) and interaction terms.
- Interaction terms capture the combined effects of variables, potentially improving the model’s ability to explain `HP`.
- We avoided extremely large interaction combinations to prevent excessive computational demands and overfitting.

**Purpose:**
The goal was to see if adding complexity improves the in-sample R-squared, while subsequent out-of-sample testing would reveal if the increased complexity leads to overfitting.

#### Cell 5
**Code Explanation:**
Finally, we calculated in-sample and out-of-sample R-squared values for the more complex model, as in Cell 3.
- We discussed that a high in-sample but low out-of-sample R-squared would indicate overfitting.
- If these values are similar, it would suggest that the complexity added is capturing meaningful patterns rather than noise.

**Purpose:**
This step confirms whether the complex model generalizes better or merely overfits the training data.

---

#### Overall Purpose of the Five Cells

Across these exchanges, we covered the process of:
1. **Training and testing with a simple model**, establishing a baseline.
2. **Evaluating generalizability** by comparing in-sample and out-of-sample R-squared values.
3. **Adding complexity** to see if it enhances or hinders generalizability.
4. **Comparing models** to determine if increased complexity truly captures meaningful relationships or simply overfits.


### 6. Work with a ChatBot to understand how the *model4_linear_form* (*linear form* specification of  *model4*) creates new *predictor variables* as the columns of the so-called "design matrix" *model4_spec.exog* (*model4_spec.exog.shape*) used to predict the *outcome variable*  *model4_spec.endog* and why the so-called *multicollinearity* in this "design matrix" (observed in *np.corrcoef(model4_spec.exog)*) contribues to the lack of "out of sample" *generalization* of *predictions* from *model4_fit*; then, explain this consisely in your own works

In [8]:
# "Cond. No." WAS 343.0 WITHOUT to centering and scaling
model3_fit.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        21:30:05   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [9]:
from patsy import center, scale

model3_linear_form_center_scale = \
  'HP ~ scale(center(Attack)) + scale(center(Defense))' 
model_spec3_center_scale = smf.ols(formula=model3_linear_form_center_scale,
                                   data=pokeaman_train)
model3_center_scale_fit = model_spec3_center_scale.fit()
model3_center_scale_fit.summary()
# "Cond. No." is NOW 1.66 due to centering and scaling

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        21:30:05   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 69.3025      1.186     58.439      0.000      66.971      71.634
scale(center(Attack))      8.1099      1.340      6.051      0.000       5.475      10.745
scale(center(Defense))     2.9496      1.340      2.201      0.028       0.315       5.585
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         1.66
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [10]:
model4_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Defense))'
model4_linear_form_CS += ' * scale(center(Speed)) * Legendary' 
model4_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# Legendary is an indicator, so we don't center and scale that

model4_CS_spec = smf.ols(formula=model4_linear_form_CS, data=pokeaman_train)
model4_CS_fit = model4_CS_spec.fit()
model4_CS_fit.summary().tables[-1]  # Cond. No. is 2,250,000,000,000,000

# The condition number is still bad even after centering and scaling


<class 'statsmodels.iolib.table.SimpleTable'>

In [11]:
# Just as the condition number was very bad to start with
model4_fit.summary().tables[-1]  # Cond. No. is 12,000,000,000,000,000

<class 'statsmodels.iolib.table.SimpleTable'>

#### How the model4_linear_form (linear form specification of model4) creates new predictor variables as the columns of the so-called "design matrix" model4_spec.exog (model4_spec.exog.shape) used to predict the outcome variable model4_spec.endog
In model4_linear_form, we’re creating a model to predict the outcome variable "HP" using several predictor variables (Attack, Defense, Speed, Legendary, Sp. Def, and Sp. Atk), but with a twist: we’re including all possible interaction terms among them. These interaction terms allow the model to consider combined effects between variables, so instead of just seeing how each individual stat (like Attack or Defense) affects HP, it can also analyze how pairs or groups of stats interact to influence HP. This setup generates a large number of new predictors, which are included as columns in what’s called the model's "design matrix" (model4_spec.exog).

The design matrix is essentially a big table where each column is one of these predictor variables or interaction terms, and each row represents a data point. The shape of this matrix (model4_spec.exog.shape) shows how many predictors we’ve created and how many observations we’re working with. In short, the design matrix is where all our predictors—both individual variables and interaction terms—are set up as inputs that the model uses to make predictions.

#### Why the so-called multicollinearity in this "design matrix" (observed in np.corrcoef(model4_spec.exog)) contribues to the lack of "out of sample" generalization of predictions from model4_fit
Multicollinearity occurs when two or more predictor variables in a model are highly correlated, meaning they share a lot of overlapping information. In our case, the design matrix for model4 (which we can inspect with np.corrcoef(model4_spec.exog)) shows that many predictors are highly correlated due to the inclusion of so many interaction terms. This excessive overlap between predictors makes it harder for the model to isolate the effect of each variable, leading to unstable estimates. As a result, the model can pick up on specific, random patterns in the training data rather than real, generalizable trends.

This is why multicollinearity contributes to poor out-of-sample performance. When we test the model on new data, it struggles to perform well because the patterns it "learned" in the training data are more noise than meaningful signals. The model has essentially been overfit—it’s too tailored to the specifics of the training data and can’t generalize well to other data. So, multicollinearity, driven by the model's high complexity, has a negative impact on generalizability, resulting in a model that works well on the training set but performs poorly on unseen data.

---
### Summaries of my ChatBot session (Q#6) (04)
**Link to the this chat log history for "Pre-lecture" HW:** https://chatgpt.com/share/67338e38-8c40-8011-b893-13b51611c920

**Summary of Interaction on Model Specification, Design Matrix, and Multicollinearity**

1. **Clarification of the Question**: We began by clarifying the assignment prompt, which asked about understanding the creation of new predictor variables in the design matrix (`model4_spec.exog`) through the linear form specification (`model4_linear_form`) and examining how multicollinearity affects model generalization in `model4`.

2. **Explanation of Linear Form Specification**: I explained that the **linear form specification** is the formula that defines which predictors, interactions, and transformations are included in the regression model. This specification generates columns in the design matrix, each representing a predictor or interaction term that the model uses for fitting.

3. **Description of Multicollinearity**: I provided a formal and causal explanation of **multicollinearity**:
   - Formally, it’s the presence of high correlations between predictor variables, leading to unstable coefficient estimates and inflated variances.
   - Causally, it arises from overlapping information between predictors, which can be due to redundancy, shared underlying factors, or accidental correlation in smaller datasets.

4. **Structure of the Design Matrix**: I illustrated how the design matrix organizes predictors and interactions as columns, with each row representing an observation. We went over several examples, showing basic design matrices, interaction terms, and transformations like centering and scaling.

5. **Answering Specific Questions**:
   - **(a)** How `model4_linear_form` creates new predictor variables in the design matrix: I explained that `model4_linear_form` generates the design matrix by expanding the predictors, interactions, and transformations specified. Each unique form or combination appears as a separate column, allowing the model to capture both individual and combined effects of predictors.
   - **(b)** Why multicollinearity in this design matrix affects out-of-sample generalization: We discussed how multicollinearity leads to unstable coefficients that fit noise in the training data rather than meaningful patterns. This causes overfitting, meaning the model performs well on the training data but poorly on new data due to capturing non-generalizable patterns.

This summary captures our main points: understanding how the design matrix is constructed, recognizing the impact of multicollinearity, and explaining why this leads to overfitting and poor generalization in `model4`.


---
### 7. Discuss with a ChatBot the rationale and principles by which *model5_linear_form* is  extended and developed from *model3_fit* and *model4_fit*; *model6_linear_form* is  extended and developed from *model5_linear_form*; and *model7_linear_form* is  extended and developed from *model6_linear_form*; then, explain this breifly and consisely in your own words

In [12]:
# Here's something a little more reasonable...
model5_linear_form = 'HP ~ Attack + Defense + Speed + Legendary'
model5_linear_form += ' + Q("Sp. Def") + Q("Sp. Atk")'
model5_linear_form += ' + C(Generation) + C(Q("Type 1")) + C(Q("Type 2"))'

model5_spec = smf.ols(formula=model5_linear_form, data=pokeaman_train)
model5_fit = model5_spec.fit()
model5_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.392
Model:                            OLS   Adj. R-squared:                  0.313
Method:                 Least Squares   F-statistic:                     4.948
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           9.48e-19
Time:                        21:30:05   Log-Likelihood:                -1765.0
No. Observations:                 400   AIC:                             3624.
Df Residuals:                     353   BIC:                             3812.
Df Model:                          46                                         
Covariance Type:            nonrobust                                         
==============================================================================================
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
Intercept                     10.1046     14.957      0.676      0.500     -19.312      39.521
Legendary[T.True]             -3.2717      4.943     -0.662      0.508     -12.992       6.449
C(Generation)[T.2]             9.2938      4.015      2.315      0.021       1.398      17.189
C(Generation)[T.3]             2.3150      3.915      0.591      0.555      -5.385      10.015
C(Generation)[T.4]             4.8353      4.149      1.165      0.245      -3.325      12.995
C(Generation)[T.5]            11.4838      3.960      2.900      0.004       3.696      19.272
C(Generation)[T.6]             4.9206      4.746      1.037      0.300      -4.413      14.254
C(Q("Type 1"))[T.Dark]        -1.4155      6.936     -0.204      0.838     -15.057      12.226
C(Q("Type 1"))[T.Dragon]       0.8509      6.900      0.123      0.902     -12.720      14.422
C(Q("Type 1"))[T.Electric]    -6.3641      6.537     -0.974      0.331     -19.220       6.491
C(Q("Type 1"))[T.Fairy]       -1.9486     10.124     -0.192      0.847     -21.859      17.962
C(Q("Type 1"))[T.Fighting]     7.0308      7.432      0.946      0.345      -7.586      21.648
C(Q("Type 1"))[T.Fire]         3.0779      6.677      0.461      0.645     -10.055      16.210
C(Q("Type 1"))[T.Flying]      -2.1231     22.322     -0.095      0.924     -46.025      41.779
C(Q("Type 1"))[T.Ghost]        5.7343      8.488      0.676      0.500     -10.960      22.429
C(Q("Type 1"))[T.Grass]        3.3275      5.496      0.605      0.545      -7.481      14.136
C(Q("Type 1"))[T.Ground]       9.5118      7.076      1.344      0.180      -4.404      23.428
C(Q("Type 1"))[T.Ice]         -0.9313      7.717     -0.121      0.904     -16.108      14.246
C(Q("Type 1"))[T.Normal]      18.4816      5.312      3.479      0.001       8.034      28.929
C(Q("Type 1"))[T.Poison]       8.3411      7.735      1.078      0.282      -6.871      23.554
C(Q("Type 1"))[T.Psychic]      1.8061      6.164      0.293      0.770     -10.317      13.930
C(Q("Type 1"))[T.Rock]        -3.8558      6.503     -0.593      0.554     -16.645       8.933
C(Q("Type 1"))[T.Steel]       -4.0053      8.044     -0.498      0.619     -19.826      11.816
C(Q("Type 1"))[T.Water]        9.7988      5.166      1.897      0.059      -0.361      19.959
C(Q("Type 2"))[T.Dark]         5.8719     15.185      0.387      0.699     -23.993      35.737
C(Q("Type 2"))[T.Dragon]      13.2777     14.895      0.891      0.373     -16.016      42.571
C(Q("Type 2"))[T.Electric]    14.3228     17.314      0.827      0.409     -19.728      48.374
C(Q("Type 2"))[T.Fairy]        2.8426     14.268      0.199      0.842     -25.218      30.903
C(Q("Type 2"))[T.Fighting]     1.9741     14.089      0.140      0.889     -25.735      29.683
C(Q("Type 2"))[T.Fire]         0.2001     1

In [13]:
yhat_model5 = model5_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model5_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model5)[0,1]**2)

'In sample' R-squared:     0.3920134083531893
'Out of sample' R-squared: 0.30015614488652215


In moving from model3_fit and model4_fit to model5_linear_form, the goal was to build a model that’s both interpretable and effective at predicting HP.

model3_fit was very simple, using only Attack and Defense, and explained little variance in HP (R-squared = 0.148). model4_fit added more variables and many interaction terms, increasing the R-squared to 0.467 but making the model complex and prone to multicollinearity (with a condition number of 1.2e+16).

model5_linear_form strikes a balance by keeping essential predictors (Attack, Defense, Speed, Legendary, Sp. Def, Sp. Atk) and adding broader categories like Generation and Type without interactions. This simplified approach reduces multicollinearity and keeps the model interpretable, achieving an R-squared of 0.392—sacrificing some predictive power for better stability and generalizability.

---

In [14]:
# Here's something a little more reasonable...
model6_linear_form = 'HP ~ Attack + Speed + Q("Sp. Def") + Q("Sp. Atk")'
# And here we'll add the significant indicators from the previous model
# https://chatgpt.com/share/81ab88df-4f07-49f9-a44a-de0cfd89c67c
model6_linear_form += ' + I(Q("Type 1")=="Normal")'
model6_linear_form += ' + I(Q("Type 1")=="Water")'
model6_linear_form += ' + I(Generation==2)'
model6_linear_form += ' + I(Generation==5)'

model6_spec = smf.ols(formula=model6_linear_form, data=pokeaman_train)
model6_fit = model6_spec.fit()
model6_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.333
Model:                            OLS   Adj. R-squared:                  0.319
Method:                 Least Squares   F-statistic:                     24.36
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           2.25e-30
Time:                        21:30:05   Log-Likelihood:                -1783.6
No. Observations:                 400   AIC:                             3585.
Df Residuals:                     391   BIC:                             3621.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
Intercept                             22.8587      3.876      5.897      0.000      15.238      30.479
I(Q("Type 1") == "Normal")[T.True]    17.5594      3.339      5.258      0.000      10.994      24.125
I(Q("Type 1") == "Water")[T.True]      9.0301      3.172      2.847      0.005       2.794      15.266
I(Generation == 2)[T.True]             6.5293      2.949      2.214      0.027       0.732      12.327
I(Generation == 5)[T.True]             8.4406      2.711      3.114      0.002       3.112      13.770
Attack                                 0.2454      0.037      6.639      0.000       0.173       0.318
Speed                                 -0.1370      0.045     -3.028      0.003      -0.226      -0.048
Q("Sp. Def")                           0.3002      0.045      6.662      0.000       0.212       0.389
Q("Sp. Atk")                           0.1192      0.042      2.828      0.005       0.036       0.202
==============================================================================
Omnibus:                      271.290   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4238.692
Skew:                           2.651   Prob(JB):                         0.00
Kurtosis:                      18.040   Cond. No.                         618.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [15]:
yhat_model6 = model6_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model6_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2)

'In sample' R-squared:     0.3326310334310908
'Out of sample' R-squared: 0.29572460427079933


The transition from model5_linear_form to model6_linear_form focuses on simplifying the model by keeping only the most significant predictors (p-value < 0.05). In model5_linear_form, a large number of variables and categories were included, but in model6_linear_form, only the most relevant features are kept: Attack, Speed, Sp. Def, Sp. Atk, and key indicators for Type 1 (Normal, Water) and Generation (2, 5).

This streamlined approach reduces complexity, lowers multicollinearity, and improves interpretability, while maintaining the essential predictors that contribute meaningfully to HP. Although model6_linear_form has a slightly lower R-squared (0.333 vs. 0.392), it’s more stable and generalizable, focusing only on the strongest predictive relationships.

---

In [16]:
# And here's a slight change that seems to perhaps improve prediction...
model7_linear_form = 'HP ~ Attack * Speed * Q("Sp. Def") * Q("Sp. Atk")'
model7_linear_form += ' + I(Q("Type 1")=="Normal")'
model7_linear_form += ' + I(Q("Type 1")=="Water")'
model7_linear_form += ' + I(Generation==2)'
model7_linear_form += ' + I(Generation==5)'

model7_spec = smf.ols(formula=model7_linear_form, data=pokeaman_train)
model7_fit = model7_spec.fit()
model7_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.378
Model:                            OLS   Adj. R-squared:                  0.347
Method:                 Least Squares   F-statistic:                     12.16
Date:                Thu, 14 Nov 2024   Prob (F-statistic):           4.20e-29
Time:                        21:30:06   Log-Likelihood:                -1769.5
No. Observations:                 400   AIC:                             3579.
Df Residuals:                     380   BIC:                             3659.
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
==========================================================================================================
                                             coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------
Intercept                                 95.1698     34.781      2.736      0.007      26.783     163.556
I(Q("Type 1") == "Normal")[T.True]        18.3653      3.373      5.445      0.000      11.733      24.997
I(Q("Type 1") == "Water")[T.True]          9.2913      3.140      2.959      0.003       3.117      15.466
I(Generation == 2)[T.True]                 7.0711      2.950      2.397      0.017       1.271      12.871
I(Generation == 5)[T.True]                 7.8557      2.687      2.923      0.004       2.572      13.140
Attack                                    -0.6975      0.458     -1.523      0.129      -1.598       0.203
Speed                                     -1.8147      0.554     -3.274      0.001      -2.905      -0.725
Attack:Speed                               0.0189      0.007      2.882      0.004       0.006       0.032
Q("Sp. Def")                              -0.5532      0.546     -1.013      0.312      -1.627       0.521
Attack:Q("Sp. Def")                        0.0090      0.007      1.311      0.191      -0.004       0.023
Speed:Q("Sp. Def")                         0.0208      0.008      2.571      0.011       0.005       0.037
Attack:Speed:Q("Sp. Def")                 -0.0002   9.06e-05     -2.277      0.023      -0.000   -2.82e-05
Q("Sp. Atk")                              -0.7277      0.506     -1.439      0.151      -1.722       0.267
Attack:Q("Sp. Atk")                        0.0136      0.005      2.682      0.008       0.004       0.024
Speed:Q("Sp. Atk")                         0.0146      0.007      2.139      0.033       0.001       0.028
Attack:Speed:Q("Sp. Atk")                 -0.0002    5.4e-05     -3.383      0.001      -0.000   -7.65e-05
Q("Sp. Def"):Q("Sp. Atk")                  0.0103      0.007      1.516      0.130      -0.003       0.024
Attack:Q("Sp. Def"):Q("Sp. Atk")          -0.0001   6.71e-05     -2.119      0.035      -0.000   -1.03e-05
Speed:Q("Sp. Def"):Q("Sp. Atk")           -0.0002   8.82e-05     -2.075      0.039      -0.000   -9.62e-06
Attack:Speed:Q("Sp. Def"):Q("Sp. Atk")   2.03e-06   7.42e-07      2.734      0.007     5.7e-07    3.49e-06
==============================================================================
Omnibus:                      252.300   Durbin-Watson:                   1.953
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3474.611
Skew:                           2.438   Prob(JB):                         0.00
Kurtosis:                      16.590   Cond. No.                     2.34e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.34e+09. Thi

In [17]:
yhat_model7 = model7_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model7_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2)

'In sample' R-squared:     0.37818209127432456
'Out of sample' R-squared: 0.35055389205977444


The jump from model6_linear_form to model7_linear_form introduces interaction terms among the main continuous features (Attack, Speed, Sp. Def, Sp. Atk). This change allows the model to capture complex relationships between these features and their combined effects on HP, rather than treating each feature’s impact on HP as isolated.

By keeping key categorical indicators (like Type 1 and Generation) and adding these interactions, model7_linear_form improves the R-squared slightly (from 0.333 to 0.378), indicating better prediction accuracy. However, this added complexity increases multicollinearity, shown by a high condition number, which could impact model stability. Overall, model7_linear_form aims to balance capturing nuanced interactions with predictive accuracy, even if it introduces some complexity.

---

In [18]:
# And here's a slight change that seems to perhas improve prediction...
model7_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Speed))'
model7_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# We DO NOT center and scale indicator variables
model7_linear_form_CS += ' + I(Q("Type 1")=="Normal")'
model7_linear_form_CS += ' + I(Q("Type 1")=="Water")'
model7_linear_form_CS += ' + I(Generation==2)'
model7_linear_form_CS += ' + I(Generation==5)'

model7_CS_spec = smf.ols(formula=model7_linear_form_CS, data=pokeaman_train)
model7_CS_fit = model7_CS_spec.fit()
model7_CS_fit.summary().tables[-1] 
# "Cond. No." is NOW 15.4 due to centering and scaling

<class 'statsmodels.iolib.table.SimpleTable'>

In [19]:
# "Cond. No." WAS 2,340,000,000 WITHOUT to centering and scaling
model7_fit.summary().tables[-1]

<class 'statsmodels.iolib.table.SimpleTable'>

---
### Summaries of my ChatBot session (Q#7) (05)
**Link to the this chat log history for "Pre-lecture" HW:** https://chatgpt.com/share/6733ad48-2ef4-8011-b93f-803ac64c37c8

Here’s a summary of our exchanges regarding the development and rationale behind the different models you’ve been working on, suitable for a record of your homework assignment:

Our discussions focused on the progressive development of predictive models for a Pokémon dataset, specifically aimed at predicting the HP (Hit Points) of Pokémon based on various attributes. The goal was to understand how each model was built, the rationale behind selecting specific predictors and interactions, and the use of techniques to balance complexity, interpretability, and predictive power.

1. **Initial Models: `model3_fit` and `model4_fit`**
   - **`model3_fit`**: This simple baseline model predicted HP using only `Attack` and `Defense` as predictors. It was useful as a starting point but underfit the data by ignoring other potentially relevant attributes.
   - **`model4_fit`**: This model significantly increased complexity by including `Attack`, `Defense`, `Speed`, `Legendary`, `Sp. Def`, and `Sp. Atk` as predictors, with all possible high-order interactions among them. While this model captured complex relationships, it risked overfitting and introduced severe multicollinearity due to the large number of interaction terms. Additionally, the high computational load made it impractical to include categorical predictors like `Generation` and Pokémon types.

2. **Development of `model5_linear_form` from `model3_fit` and `model4_fit`**
   - **Balanced Feature Selection**: `model5` incorporated additional predictors (e.g., `Speed`, `Legendary`, `Sp. Def`, and `Sp. Atk`), which were omitted from `model3` but deemed important based on `model4`. This expanded feature set aimed to improve prediction accuracy without the risk of overfitting.
   - **Avoidance of High-Order Interactions**: Instead of using all possible interactions as in `model4`, `model5` limited itself to main effects only. This reduced multicollinearity and complexity, balancing the need for interpretability and generalizability.
   - **Inclusion of Categorical Variables**: `model5` treated `Generation`, `Type 1`, and `Type 2` as categorical variables (using `C()` in the formula). This allowed the model to capture distinct differences across Pokémon generations and types, adding nuance to the predictions while keeping the model manageable.

3. **Detailed Explanation of Model Components**
   - We discussed the purpose of each term in `model5`, explaining why `Q()` was used for variables with special characters (e.g., `"Sp. Def"` and `"Sp. Atk"`) and `C()` was used to treat `Generation` as a categorical variable. This allowed the model to handle both unusual variable names and categorical predictors appropriately.
   
4. **Rationale and Principles of Model Evolution**
   - Our discussions highlighted the principles of evidence-based predictor selection (focusing on statistically significant variables), multicollinearity control, and balancing complexity with interpretability. The evolution from `model3` to `model5` showed a thoughtful process of refining the model to include meaningful predictors while avoiding overfitting.
   - The final `model5` version was a well-balanced model that improved predictive performance and interpretability by combining selected main effects and categorical distinctions, avoiding the excessive complexity seen in `model4`.


---
### 9. Work with a ChatBot to understand the meaning of the illustration below; and, explain this in your own words

In [20]:
model7_gen1_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model7_gen1_predict_future_fit = model7_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model7_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

'In sample' R-squared:     0.37818209127432456 (original)
'Out of sample' R-squared: 0.35055389205977444 (original)
'In sample' R-squared:     0.5726118179916575 (gen1_predict_future)
'Out of sample' R-squared: 0.11151363354803218 (gen1_predict_future)


In [21]:
model7_gen1to5_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model7_gen1to5_predict_future_fit = model7_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model7_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

'In sample' R-squared:     0.37818209127432456 (original)
'Out of sample' R-squared: 0.35055389205977444 (original)
'In sample' R-squared:     0.3904756578094535 (gen1to5_predict_future)
'Out of sample' R-squared: 0.23394915464343125 (gen1to5_predict_future)


In [22]:
model6_gen1_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model6_gen1_predict_future_fit = model6_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model6_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

'In sample' R-squared:     0.3326310334310908 (original)
'Out of sample' R-squared: 0.29572460427079933 (original)
'In sample' R-squared:     0.4433880517727282 (gen1_predict_future)
'Out of sample' R-squared: 0.1932858534276128 (gen1_predict_future)


In [23]:
model6_gen1to5_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model6_gen1to5_predict_future_fit = model6_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model6_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

'In sample' R-squared:     0.3326310334310908 (original)
'Out of sample' R-squared: 0.29572460427079933 (original)
'In sample' R-squared:     0.33517279824114776 (gen1to5_predict_future)
'Out of sample' R-squared: 0.26262690178799936 (gen1to5_predict_future)


The illustration is a demonstration of how training data selection (using either a single generation or multiple generations) affects model performance across two different models—model6 and model7—when used to predict outcomes for data outside the training subset. In this case, we apply the models under two different conditions: (a) training on generation 1 data only and (b) training on generations 1 through 5, excluding generation 6. These conditions allow us to explore how each model performs within the context of both "in-sample" predictions (predictions within the data subset used to fit the model) and "out-of-sample" predictions (predictions on data outside that subset).

When we train both models on generation 1 only, we observe that both models perform relatively well on this specific generation, with model7 achieving a higher in-sample R-squared than model6. This indicates that model7 is better at capturing the variability within generation 1 data, likely due to its added complexity, which includes multiple interaction terms. However, when we apply these models to predict outcomes on data outside of generation 1—namely, generations 2-6—their out-of-sample performance shows a significant decline. The drop is much sharper for model7 than for model6, signaling that model7 may be overfitting to patterns that are unique to generation 1 and do not generalize well to other generations. Model6, with its simpler structure, maintains a more modest and consistent out-of-sample R-squared, demonstrating a better generalizability despite its lower in-sample fit.

In the second condition, where we train on generations 1 through 5, excluding generation 6, we see a different pattern. Both models again achieve decent in-sample R-squared values, but this time their training data includes a broader range of generations, which introduces more variability and potentially strengthens generalizable patterns within the model. When these models are applied to predict outcomes for generation 6, we again observe that model7 achieves a higher in-sample R-squared. However, when tested on out-of-sample data (generation 6), model7's performance, though somewhat better than in the first condition, still exhibits instability and a noticeable drop in R-squared, indicating a tendency to overfit to the broader but still limited training set (generations 1-5). Model6, in contrast, maintains a stable out-of-sample R-squared with generation 6, albeit with a slightly lower in-sample R-squared than model7. This consistency further suggests that model6 is better suited to generalizing across generations without overfitting to specific generation-specific details in the training data.

---
### Summaries of my ChatBot session (Q#9) (06)
**Link to the this chat log history for "Pre-lecture" HW:** https://chatgpt.com/share/67368071-9038-8011-b74e-526a4daee06e

#### Analysis and Interpretation of Code Cells (1–4) for Pokémon Model Generalizability

#### Overview:
The assignment focused on understanding the generalizability and interpretability of two linear regression models (`model7` and `model6`) when trained on Pokémon data split by generation. We examined in-sample and out-of-sample R-squared values to evaluate how well each model performed on both the training data and new, unseen data from future generations.

---

#### Code Cell 1:
- **Objective**: Fit the complex `model7` model using only Generation 1 data, then evaluate its ability to generalize to other generations.
- **Findings**: The Gen1-only `model7` achieved a high in-sample R-squared of `0.5726` but a significantly lower out-of-sample R-squared of `0.1115`. This drop indicated poor generalizability, suggesting `model7` overfit to Generation 1 data and struggled with other generations.
  
#### Code Cell 2:
- **Objective**: Fit `model7` using data from Generations 1 to 5, then evaluate its ability to generalize to Generation 6.
- **Findings**: The Gen1-5 `model7` had an in-sample R-squared of `0.3905` and an improved out-of-sample R-squared of `0.2339` on Generation 6. The increased out-of-sample performance demonstrated that training on a broader dataset enhanced generalizability, though `model7` still showed some signs of overfitting.

---

#### Code Cell 3:
- **Objective**: Fit the simpler `model6` model using only Generation 1 data, then evaluate its generalizability on other generations.
- **Findings**: The Gen1-only `model6` achieved an in-sample R-squared of `0.4434` and an out-of-sample R-squared of `0.1933`. Although `model6` also showed reduced out-of-sample performance, it maintained better generalizability than `model7` when trained on limited data (Generation 1 only).

#### Code Cell 4:
- **Objective**: Fit `model6` using data from Generations 1 to 5, then evaluate its ability to generalize to Generation 6.
- **Findings**: The Gen1-5 `model6` achieved an in-sample R-squared of `0.3352` and an out-of-sample R-squared of `0.2626`. This model displayed both consistent in-sample performance and improved out-of-sample generalizability compared to the Gen1-only version, reinforcing that training on a broader dataset improves predictions on future data.

---

#### Key Insights and Conclusions:

1. **Model Complexity vs. Generalizability**:
   - `model7` (complex model) showed signs of overfitting, particularly when trained on limited data (Gen1-only), as evidenced by high in-sample but low out-of-sample R-squared values.
   - `model6` (simpler model) maintained better generalizability with more consistent out-of-sample performance, making it more suitable for generalization tasks.

2. **Impact of Expanded Training Data**:
   - Training on multiple generations (Gen1-5) improved both models’ out-of-sample performance, confirming that a diverse training dataset better captures generalizable patterns.

3. **Interpretability and Model Selection**:
   - Despite `model7`’s marginally higher performance, `model6` was preferable due to its simplicity, interpretability, and stable performance across generations.
   - A simpler model (`model6`) generally offers a better balance between interpretability and generalizability, which is advantageous for tasks requiring predictions on unseen data.

4. **Recommendation**:
   - For predictive tasks involving future generations, `model6` trained on a broad dataset (Gen1-5) is recommended, as it offers a reliable trade-off between interpretability and generalization.

---

This analysis highlighted the benefits of model parsimony and reinforced the importance of generalizability over complexity, especially when models are applied to sequential or generational data.